### PR TITLE
yarn SBOM generation plugin

### PR DESCRIPTION
**Pull request by AugustusKling**
This contains a working draft of a Yarn plugin to generate SBOMs. Please see README.md for usage instructions.

Note that I'm not too familiar with the CycloneDX file format and things might be off even though validators pass on the generated files.

Open things:
* What are sensible defaults for the parameters?
* Decision on license. Am I correct to assume the contained components can be used as part of a Apache-2.0 licensed project? Are you aware of any tool that takes a list of SPDX identifiers for dependencies (or SBOM) and answers which licenses could be chosen?
* Treatment of the root component in the SBOM. Should it be put in the `metadata` only or also within the `components` property?
* Many files in the root folder are basically empty templates.
* Do you have sample projects with known expected output regarding listed components and licenses?
* Fallback for component licenses.
* How would one calculate the hash for https://cyclonedx.org/docs/1.5/json/#components_items_hashes for a folder representing a component?
* Which are the important/desired properties in the generated SBOM files besides those marked as required in the JSON schema?
* Is the reproducible results flag mentioned in #8 important? Shouldn't this be a task for a SBOM consumer by masking out properties?
* I've only used this plugin with Yarn 4.1.0. It may or may not work with other Yarn versions.

Please play around with the draft plugin and give feedback.

### CODE REVIEW

File: generate-components-licenses.mjs [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/**
+ * This script is intended to be used with this repository only.
```
Comment (AugustusKling) [inline generate-components-licenses.mjs:2]: This is not part of the compiled Yarn plugin bundle. It only serves to generate `bundles/components-licenses.md`

File: package.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "yarn-plugin-sbom",
+  "version": "1.0-dev",
+  "license": "Apache-2.0",
```
Comment (AugustusKling) [inline package.json:4]: Up for discussion, see comment in PR description.

File: sources/sbom.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+  const bom = new CDX.Models.Bom();
+  bom.metadata.timestamp = new Date();
+
+  const allDependencies = await traverseWorkspace(project, workspace, config);
+  const componentModels = new Map<LocatorHash, CDX.Models.Component>();
+  // Build models without their dependencies.
+  for (const pkgInfo of allDependencies) {
+    const component = packageInfoToCycloneComponent(pkgInfo);
+    componentModels.set(pkgInfo.package.locatorHash, component);
+    bom.components.add(component);
```
Comment (AugustusKling) [inline sources/sbom.ts:47]: Should this skip over the root component referenced as `metadata.component`?

File: sources/sbom.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  }
+  if (manifest.raw.repository?.url) {
+    component.externalReferences.add(
+      new CDX.Models.ExternalReference(
+        manifest.raw.repository?.url,
+        CDX.Enums.ExternalReferenceType.VCS
+      )
+    );
+  }
+  if (manifest.license && !manifest.license.includes("SEE LICENSE")) {
```
Comment (AugustusKling) [inline sources/sbom.ts:142]: How would one refer to a license that is only found in file like `SEE LICENSE whatever.txt`?

File: sources/sbom.ts [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+  } else {
+    // TODO Handle other Yarn protocols. How to convert them?
+  }
+  return component;
+}
+
+/**
+ * Attempts to parse bogus but unambigous licenses and augments the component model.
+ */
+function attemptFallbackLicense(
```
Comment (AugustusKling) [inline sources/sbom.ts:168]: This does only set a license when outdated variants of specifying licenses were found. It does not guess from license texts in files as I'm not confident that this would guess licenses correctly.

File: sources/traverseUtils.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+};
+
+const fileNameOptions = ["license", "licence", "unlicense", "unlicence"];
+const fileNameOptionsStart = fileNameOptions.map((name) => name + ".");
+
+function readLicenseFile(
+  packageRoot: PortablePath,
+  packageFs: FakeFS<PortablePath>
+): string {
+  const files = packageFs.readdirSync(packageRoot).filter((f) => {
```
Comment (AugustusKling) [inline sources/traverseUtils.ts:114]: Calling `packageFs.readdirSync(...)` feels wrong but appears to work. I couldn't find API documentation for Yarn to clarify how their API should be used.

**Comment by jkowalleck** [comment]
thank you for the PR. I see some open questions from your side. 

we could discuss those in here: https://github.com/CycloneDX/cyclonedx-node-yarn/discussions/14 
I've transferred your current questions over there, and started answering some, wil answer the others soon.
Feel free to add more questions, or respond to my answers.

File: sources/sbom.ts (hunk 4 above)
Comment (jkowalleck) [reply sources/sbom.ts:142]: I'd add no special handing for this.
if the license text is `"SEE LICENSE whatever.txt"` then I'd also expect the `pkgInfo.licenseFileContent` to point to that file - which would cause your implementation to include the text.

File: sources/sbom.ts (hunk 5 above)
Comment (jkowalleck) [reply sources/sbom.ts:168]: if not confident, just do not add any licenses to the SBOM. 
better have an "incomplete" SBOM, than a wrong SBOM.

File: sources/sbom.ts [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    );
+  }
+  if (manifest.license && !manifest.license.includes("SEE LICENSE")) {
+    const license = licenseFactory.makeFromString(manifest.license);
+    if (
+      pkgInfo.licenseFileContent &&
+      (license instanceof CDX.Models.NamedLicense ||
+        license instanceof CDX.Models.SpdxLicense)
+    ) {
+      license.text = new CDX.Models.Attachment(pkgInfo.licenseFileContent);
```
Comment (jkowalleck) [inline sources/sbom.ts:149]: please remove the whole license gathering process, for now.  Tis feature should not be part of a minimum viable product. 
It might be introduced in a feature development in the future.

Gathering license information is the area where lawyers act strange.
Better have an incomplete SBOM than a wrong SBOM.


gathering license texts is a feature that should be enabled by a CLI switch. Per default, no license texts should be gathered, as this causes SBOM bloat.

File: sources/sbom.ts [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    component.licenses.add(license);
+  } else {
+    attemptFallbackLicense(manifest, pkg, component);
+  }
+
+  const devirtualizedLocator = structUtils.ensureDevirtualizedLocator(pkg);
+  if (devirtualizedLocator.reference.startsWith("npm:")) {
+    component.purl = npmPurlFactory.makeFromComponent(component);
+  } else {
+    // TODO Handle other Yarn protocols. How to convert them?
```
Comment (jkowalleck) [inline sources/sbom.ts:160]: could you list these protocols in the `TODO` comment, so that any developer knows about expected features?

File: sources/sbom.ts [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+    }
+  }
+}
+
+/**
+ * Converts CycloneDX XML to xmlbuilder2 XML models.
+ * @param parent Model of xmlbuilder2
+ * @param e Model of CycloneDX
+ */
+function convertXML(
```
Comment (jkowalleck) [inline sources/sbom.ts:220]: why is this needed? the CycloneDX library ships with a complete XML serializer!
example usage: https://github.com/CycloneDX/cyclonedx-javascript-library/blob/f8d52796e3645cc0bc7308f50fc0973b2d982e49/examples/node/typescript/example.mjs/src/example.ts#L71-L73

File: sources/sbom.ts [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+    }
+  }
+}
+
+type SupportedSpec =
+  | typeof CDX.Spec.Spec1dot2
+  | typeof CDX.Spec.Spec1dot3
+  | typeof CDX.Spec.Spec1dot4
+  | typeof CDX.Spec.Spec1dot5;
+function spec(specVersion: CDX.Spec.Version): SupportedSpec {
```
Comment (jkowalleck) [inline sources/sbom.ts:248]: Why is this needed? 
Could  use `CDX.Spec.SpecVersionDict` instead? see https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/src/spec/consts.ts#L308

File: sources/sbom.ts [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+      return serializer.serialize(bom, {
+        space: 2,
+      });
+    }
+    case CDX.Spec.Format.XML: {
+      const serializer = new CDX.Serialize.XmlSerializer(
+        new CDX.Serialize.XML.Normalize.Factory(serializeSpec)
+      );
+      const e = serializer.normalizerFactory.makeForBom().normalize(bom, {});
+      // CycloneDX's XML serializer cannot find its xmlbuilder2 dependency, at least when bundled as Yarn plugin.
```
Comment (jkowalleck) [inline sources/sbom.ts:98]: > CycloneDX's XML serializer cannot find its xmlbuilder2 dependency, at least when bundled as Yarn plugin.

why is that? could you investigate and help fix this issue (maybe upstream in https://github.com/CycloneDX/cyclonedx-javascript-library)?

**Comment by jkowalleck** [comment]
@AugustusKling in order to use your work, i need you to sign the DCO.
implications: https://github.com/CycloneDX/.github/blob/master/CONTRIBUTING.md#contributing-code-and-pull-requests

 see a how to here: https://github.com/CycloneDX/cyclonedx-node-yarn/pull/13/checks?check_run_id=21791053467

File: .gitignore
[the remark this answers is not in the document]
Comment (AugustusKling) [reply .gitignore:1]: Most of the content of the `.gitignore` was not relevant. I've added it back except for the Yarn things with follow the proposal fromhttps://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored

File: README.md [hunk 12]
```diff
@@ -1,11 +1,12 @@
-# cyclonedx-node-yarn
```
[the remark this answers is not in the document]
Comment (AugustusKling) [reply README.md:1]: Yarn's plugin builder has a naming pattern and refuses to compile otherwise.
`Usage Error: Invalid plugin name "cyclonedx-node-yarn" - it should be "yarn-plugin-<something>"`

Please suggest some other name that matches this pattern.

File: bundles/@yarnpkg/plugin-sbom.js
[the remark this answers is not in the document]
Comment (AugustusKling) [reply bundles/@yarnpkg/plugin-sbom.js:1]: Yes, it's the build result. I've put it in as I've seen it done likewise for other Yarn plugins.

Keeping it like this is not important as long as the build result is 1 file and accessible using ordinary HTTPS somewhere. This file is imported by https://yarnpkg.com/cli/plugin/import by the people wanting to use it without building themselves.

File: sources/sbom.ts (hunk 7 above)
Comment (AugustusKling) [reply sources/sbom.ts:189]: I'd prefer to have license info available in the SBOMs as I intend to use it to verify licenses of transitve dependencies (in a bunch of non-public projects).

The other use is the generation of `./bundles/components-licenses.md` which is required per the dependecies of this Yarn plugin as long as a bundle is provided.

Adding the license into is no longer a default thing, though. In the meantime you need to enable it with `--licenses` if desired.

File: sources/sbom.ts (hunk 10 above)
Comment (AugustusKling) [reply sources/sbom.ts:280]: `CDX.Spec.SpecVersionDict` has its key type defined as `CDX.Spec.Version` which has more values than contained in the dictionary. One could extract the spec values like the following but to me listing the supported values is easier to read and gives a more precise type. Either way would work, though.
```
type ValueOf<T> = T[keyof T];
type SupportedSpec = ValueOf<typeof CDX.Spec.SpecVersionDict>;
```

File: sources/sbom.ts (hunk 11 above)
Comment (AugustusKling) [reply sources/sbom.ts:115]: Probably in @cyclonedx/cyclonedx-library/libs/universal-node-xml/index.js where it says
```const possibleStringifier = require(`./stringifiers/${file}`)```

Yarn's plugin bundler compiles to 1 file and my guess is that it fails to handle such runtime `require` calls.

File: sources/sbom.ts (hunk 9 above)
Comment (AugustusKling) [reply sources/sbom.ts:252]: See https://github.com/CycloneDX/cyclonedx-node-yarn/pull/13#discussion_r1498362383

**Comment by AugustusKling** [comment]
@jkowalleck thank you very much for your feedback so far. It's much appreciated to get your prespective on this.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jkowalleck) [reply README.md:1]: please choose a different headlnie, then.
something along with "CycloneDX SBOM plugin for yarn"?

File: bundles/@yarnpkg/plugin-sbom.js
[the remark this answers is not in the document]
Comment (jkowalleck) [reply bundles/@yarnpkg/plugin-sbom.js:1]: Understood. like I said, we do not need to have the thing in the repo, we would have it as a release asset.

File: sources/sbom.ts (hunk 7 above)
Comment (jkowalleck) [reply sources/sbom.ts:189]: no worries, the feature is not cut entirely, but i can see that you are missing a lot of the aspects to to this feature. i dont want to go into details, as this shall be discussed in a feature request in the future.

Comment (jkowalleck) [reply sources/sbom.ts:189]: this being said, please remove the license gathering so we can have a first MVP soon,
instead of fiddling around with license gathereing.

File: sources/sbom.ts (hunk 11 above)
Comment (jkowalleck) [reply sources/sbom.ts:115]: interesting. could you check with https://github.com/CycloneDX/cyclonedx-javascript-library) and propose a process that is safe for the yarn plugin bundler?

File: sources/sbom.ts (hunk 10 above)
Comment (jkowalleck) [reply sources/sbom.ts:280]: > `CDX.Spec.SpecVersionDict` has its key type defined as `CDX.Spec.Version` which has more values than contained in the dictionary. 

This would be unexpected. Which values does the enum have, that the dict does not have?  Would you pull-request a fix for the library, so that this is no longer an issue for the usage here?

Comment (AugustusKling) [reply sources/sbom.ts:280]:
```typescript
export const SpecVersionDict: Readonly<Partial<Record<Version, Readonly<_SpecProtocol>>>> = Object.freeze({
  [Version.v1dot5]: Spec1dot5,
  [Version.v1dot4]: Spec1dot4,
  [Version.v1dot3]: Spec1dot3,
  [Version.v1dot2]: Spec1dot2
  // <= v1.1 is not implemented
})
```

And the Version enum
```typescript
export enum Version {
  v1dot5 = '1.5',
  v1dot4 = '1.4',
  v1dot3 = '1.3',
  v1dot2 = '1.2',
  v1dot1 = '1.1',
  v1dot0 = '1.0',
}
```

The versions 1.0 and 1.1 are not present in the dictionary nor are they available as protocol instances in `CDX.Spec`. But as the dictionary has its key type defined as `Version` the non-availability of versions 1.0 and 1.1 are lost.

Comment (jkowalleck) [reply sources/sbom.ts:280]: this is intentional. 

```typescript
// !!! partial
export const SpecVersionDict_1: Readonly<Partial<Record<Version, Readonly<_SpecProtocol>>>> = Object.freeze({
  [Version.v1dot5]: Spec1dot5,
  [Version.v1dot4]: Spec1dot4,
  [Version.v1dot3]: Spec1dot3,
  [Version.v1dot2]: Spec1dot2
  // <= v1.1 is not implemented
})

// !! no longer parial
export const SpecVersionDict_2: Readonly<Record<Version, Readonly<_SpecProtocol>>> = Object.freeze({
  [Version.v1dot5]: Spec1dot5,
  [Version.v1dot4]: Spec1dot4,
  [Version.v1dot3]: Spec1dot3,
  [Version.v1dot2]: Spec1dot2,
  // <= v1.1 is not implemented
  [Version.v1dot1]: undefined,
  [Version.v1dot0]: undefined
})

s1 = SpecVersionDict_1[Version.v1dot1]
s2 = SpecVersionDict_2[Version.v1dot1]
```

I'd expect `s1 == undefined` and `s2 == undefined`.
Is this not the case?

File: sources/sbom.ts (hunk 11 above)
Comment (jkowalleck) [reply sources/sbom.ts:115]: I see.
Please open an issue in [here](https://github.com/CycloneDX/cyclonedx-javascript-library/issues/new), and describe the problem (and a possible solution).
Feel free to open a PR with the intended solution. 

Your solution in the library may be pragmatic. No need to keep the current design/architecture, just pullrequest whatever would suite your needs. :D

File: bundles/@yarnpkg/plugin-sbom.js
[the remark this answers is not in the document]
Comment (AugustusKling) [reply bundles/@yarnpkg/plugin-sbom.js:1]: Removed and replace by explaination with https://github.com/CycloneDX/cyclonedx-node-yarn/pull/13/commits/0bc3fd632c5ba21da4c734ea56f25294dff617f9

File: sources/sbom.ts (hunk 10 above)
Comment (AugustusKling) [reply sources/sbom.ts:279]: Getting the `undefined` value for the 2 mentioned cases happens as expected. I've now used a type based on `CDX.Spec.SpecVersionDict` because it does not matter for my current use which specifications are supported.

File: bundles/@yarnpkg/plugin-sbom.js
[the remark this answers is not in the document]
Comment (jkowalleck) [reply bundles/@yarnpkg/plugin-sbom.js:1]: great. thank you very much

File: sources/sbom.ts (hunk 11 above)
Comment (AugustusKling) [reply sources/sbom.ts:117]: See https://github.com/CycloneDX/cyclonedx-javascript-library/issues/1017

File: package.json [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  "main": "./sources/index.ts",
+  "dependencies": {
+    "@cyclonedx/cyclonedx-library": "^6.3.1",
```
Comment (jkowalleck) [inline package.json:7]: please bump `@cyclonedx/cyclonedx-library` to `^6.3.2`

File: .vscode/extensions.json
Comment (jkowalleck) [inline .vscode/extensions.json:1]: Please remove IDE settings from the repo.
Everybody preferred a different editor/IDE and we do not intend to keep configs for everything here, especially as it would require maintenance.

you might add the `.vscode` folder to your (global) gitignore file.

File: .yarn/sdks/integrations.yml
Comment (jkowalleck) [inline .yarn/sdks/integrations.yml:1]: why is there a setting for a specific IDE? 

This seams to be nothing we need in this repo, instead, a README/instruction-set that describes and links https://yarnpkg.com/getting-started/editor-sdks would be better. 

Any obligations?

File: generate-components-licenses.mjs (hunk 1 above)
Comment (jkowalleck) [reply generate-components-licenses.mjs:2]: note for self: revisit before first beta-release happens

File: package.json (hunk 13 above)
Comment (jkowalleck) [reply package.json]: done via b8ff4f7be7be1bb9627477e9a1ff544f3659968b

File: sources/sbom.ts [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+ */
+export const stdOutOutput = Symbol();
+export interface OutputOptions {
+  specVersion: CDX.Spec.Version;
+  outputFormat: CDX.Spec.Format;
+  /** Output file name. */
+  outputFile: PortablePath | typeof stdOutOutput;
+  componentType: CDX.Enums.ComponentType;
+  /** If component licenses shall be included. */
+  licenses: boolean;
```
Comment (jkowalleck) [inline sources/sbom.ts:29]: the license gathering is probably not a output option.
either the interface name is not suitable, or the option shall be moved.

File: sources/sbom.ts [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+    const component = packageInfoToCycloneComponent(
+      pkgInfo,
+      outputOptions.licenses
+    );
+    componentModels.set(pkgInfo.package.locatorHash, component);
+    if (pkgInfo.package.locatorHash === workspace.anchoredLocator.locatorHash) {
+      // Set workspace as root component.
+      bom.metadata.component = component;
+      bom.metadata.component.type = outputOptions.componentType;
+      bom.metadata.tools.add(component);
```
Comment (jkowalleck) [inline sources/sbom.ts:66]: why add `component` as a `metadata.tool`? 

`metadata.tool` should include the SBOM generator and the `cyclonedx-javascript-library`, as they were involved in the SBOM process ...  
Tool name and version would be great, additional info is optional.

Example result like so: https://github.com/CycloneDX/cyclonedx-node-npm/blob/c7414ee0d032ba9d067001c2c27b42581b553c4b/demo/dev-dependencies/example-results/bare/bom.1.5.json#L12-L55

File: sources/sbom.ts [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+  outputFormat: OutputOptions["outputFormat"]
+): string {
+  switch (outputFormat) {
+    case CDX.Spec.Format.JSON: {
+      const serializer = new CDX.Serialize.JsonSerializer(
+        new CDX.Serialize.JSON.Normalize.Factory(serializeSpec)
+      );
+      return serializer.serialize(bom, {
+        space: 2,
+        sortLists: true,
```
Comment (jkowalleck) [inline sources/sbom.ts:109]: `sortLists: true,` is costly, it should be driven by the `outputOptions.reproducible`

File: sources/sbom.ts [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+/**
+ * @returns Model, but no dependencies set.
+ */
+function packageInfoToCycloneComponent(
+  pkgInfo: PackageInfo,
+  licenses: boolean
+): CDX.Models.Component {
+  const pkg = pkgInfo.package;
+  const manifest = pkgInfo.manifest;
+  const component = new CDX.Models.Component(
```
Comment (jkowalleck) [inline sources/sbom.ts:133]: in case you were working with normalized package manuifest data, you could utilize 
`CDX.Builders.ComponentBuilder`
see https://github.com/CycloneDX/cyclonedx-javascript-library/blob/3367c1638662d57c53caff0824ee94cd7859bee2/src/builders/fromNodePackageJson.node.ts#L70C14-L70C31

File: .yarn/sdks/integrations.yml
Comment (AugustusKling) [reply .yarn/sdks/integrations.yml:1]: No obligatations. I had followed the instructions from https://yarnpkg.com/migration/pnp#editor-support but it makes more sense if everybody does this setup according to their preferences.

File: package.json [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    "@cyclonedx/cyclonedx-library": "^6.4.0",
+    "@yarnpkg/cli": "^4.0.0",
+    "@yarnpkg/core": "^4.0.0",
+    "@yarnpkg/fslib": "^3.0.2",
+    "clipanion": "^4.0.0-rc.2",
+    "spdx-license-ids": "^3.0.17",
+    "xmlbuilder2": "^3.1.1"
+  },
+  "devDependencies": {
+    "@types/node": "^20.0.0",
```
[the remark this answers is not in the document]
Comment (AugustusKling) [reply package.json]: Do you have have a link to any documentation that explains the ts-version-tags? I had not encountered them before and would assume only the runtime environment's API (Node.js) would be relevant.

File: package.json [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    "@yarnpkg/builder": "^4.0.0",
```
[the remark this answers is not in the document]
Comment (AugustusKling) [reply package.json:17]: The lockfile `yarn.lock` has the resolved versions and ensures whover executes the `yarn install` command gets the exact same result. In fact lockfiles that do actually work was our reason to switch from NPM to Yarn a few years ago and we haven't looked back since.

Update process is to call any of Yarn's commands to update dependencies like `yarn up` or `yarn upgrade-interactive`, then verify by executing the tests that everything works.

I'd argue that the version of the builder is not important. If the version picked by the update commands generates a working plugin (meaning tests pass), the picked version is okay.

File: package.json [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    "rimraf": "5.0.0",
```
[the remark this answers is not in the document]
Comment (AugustusKling) [reply package.json]: done

File: sources/sbom.ts (hunk 15 above)
Comment (AugustusKling) [reply sources/sbom.ts]: This was an error. I've changed it to include to include `@cyclonedx/cyclonedx-library` and `yarn-plugin-sbom` instead.

What I need to think about is a lean way to get their versions as part of the build process.

What we need to solve in `@cyclonedx/cyclonedx-library` is to generate correct output when receiving `CDX.Models.Component` instances as tools. The schema at https://cyclonedx.org/docs/1.5/json/#metadata_tools wants ordinary components over the legacy creation tools, which is good, but the `ToolRepository` in the library does not handle components. This is the reason for the missing group `@cyclonedx` in the generated SBOM files.

File: sources/sbom.ts (hunk 16 above)
Comment (AugustusKling) [reply sources/sbom.ts]: done

File: package.json (hunk 18 above)
[the remark this answers is not in the document]
Comment (jkowalleck) [reply package.json]: it is just the recommended way to keep your TS in sync with the latest compatible typeings.

File: sources/sbom.ts (hunk 15 above)
Comment (jkowalleck) [reply sources/sbom.ts]: > What I need to think about is a lean way to get their versions as part of the build process. 

👍 

> The schema at https://cyclonedx.org/docs/1.5/json/#metadata_tools wants ordinary components over the legacy creation tools, 

True. For backwards compatibility with older versions of CycloneDX, you should go with `Tool` for now.

File: sources/sbom.ts [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+    CDX.Enums.ComponentType.Library,
+    pkg.name,
+    {
+      bomRef: pkgInfo.package.locatorHash,
+      group: pkg.scope ? `@${pkg.scope}` : undefined,
+      version: packageVersionWithManifestFallback(pkg, manifest),
+      author: manifest.raw.author?.name,
```
[the remark this answers is not in the document]
Comment (AugustusKling) [reply sources/sbom.ts]: Good point. Support for string authors was not there but is there, now.

The library looks interesting for other purposes but I'd like to keep the number of dependencies including transitive dependencies low. Also, it copies a lot of data around with is not necessary to just extract the author name.

I tried to implement it according to https://docs.npmjs.com/cli/v10/configuring-npm/package-json#people-fields-author-contributors and verified that my implementation produces the same output than `normalize-package-data`.

File: sources/sbom.ts (hunk 17 above)
Comment (AugustusKling) [reply sources/sbom.ts]: This works nicely.

File: sources/sbom.ts [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+      description: manifest.raw.description,
+    }
+  );
+  if (manifest.raw.homepage) {
+    component.externalReferences.add(
+      new CDX.Models.ExternalReference(
```
[the remark this answers is not in the document]
Comment (AugustusKling) [reply sources/sbom.ts]: Done by using in `CDX.Builders.ComponentBuilder` in the meantime.

File: sources/sbom.ts (hunk 15 above)
Comment (AugustusKling) [reply sources/sbom.ts]: I switched to the older `CDX.Models.Tool` and also added their version numbers.

**Comment by jkowalleck** [comment]
your work looks very promising, @AugustusKling .

Before we can move forward, I need you to do the following:
Please sign off your commits, to show that you agree to publish your changes under the current terms and licenses of the project , and to indicate agreement with [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

An explanation how to sign off after commits were already pushed can always be found in the pull request's check "DCO" details.
For example here: https://github.com/CycloneDX/cyclonedx-node-yarn/pull/13/checks?check_run_id=22140795320

**Comment by AugustusKling** [comment]
@jkowalleck, I had not done the sign-off yet because I feared the rebase to do so would create a mess with the comments in this pull request. It seems through, that rebasing did not create much trouble and edited the commit messages to put the desired sign-off line.

File: sources/sbom.ts (hunk 11 above)
Comment (jkowalleck) [reply sources/sbom.ts]: appears to be done will close this very conversation as "resolved"
:+1:

**Comment by jkowalleck** [comment]
@AugustusKling would you be okay if i merged the current sattus?
It all looks good, and I woudl like to continue development from the current satte.

my next steps would be to setup some QA/CT pipelines, polish or modify some features here and there, revisit the test cases, enhance the documentation, and write additional feature requests (requirements engineering) for further development.
Until the licensing question for the release publication is finally clarified with OWASP (in progress), there is still some thingfs to work on.

If you wanted to continue contribution, I coult ping you as soon as new updates come along.
Is this something you would be interested in?

**Comment by AugustusKling** [comment]
> @AugustusKling would you be okay if I merged the current states? It all looks good, and I would like to continue development from the current state.

@jkowalleck, I wanted to make the initial build simpler before merging which I did earlier today. Feel free to go ahead with the merge, now.


> If you wanted to continue contribution, I could ping you as soon as new updates come along. Is this something you would be interested in?

Gladly, please inform me about your plans.

I wanted to take the generated SBOM to verify only desired licenses are present in projects, including its transitive dependencies. It seems the generated SBOM is good enough for this purpose. The packages for which the code fails to detect the license (as in license information is absent from the SBOM) can be handled manually as not so many affected packages remain. Not in the projects I have tested with anyway.

The other thing I have in mind is creating a tool which takes the SBOM files, looks for undesired packages in the dependeny tree, and reports these. Undesired is a flexible term here as the criteria would depend on the user. What I have in mind is scanning for packages with known vulnerabilities, packages with incompatible licenses, packages which are not updated and those where their release date is not in a given time range.

These uses would just consume the SBOM but exist outwith the scope of this Yarn plugin.

**Review by jkowalleck — APPROVED** [review]

**Comment by jkowalleck** [comment]
> I wanted to take the generated SBOM to verify only desired licenses are present in projects, including its transitive dependencies. [...]
> [...]
> The other thing I have in mind is creating a tool which takes the SBOM files, looks for undesired packages in the dependeny tree, and reports these. [...] scanning for packages with known vulnerabilities, packages with incompatible licenses, packages which are not updated and those where their release date is not in a given time range.

For example, CycloneDX' sister project - [Dependency Track](https://dependencytrack.org/), can ingest an SBOM and do all the things you mentioned.
You could define several profiles to prevent and detect unwanted packages, licenses, and so on.
And it tells about outdated and vulnerable packages.

**Comment by jkowalleck** [comment]
re: https://github.com/CycloneDX/cyclonedx-node-yarn/pull/13#issuecomment-1975160143

> Gladly, please inform me about your plans.

will do, @AugustusKling :+1:
